php/libraries/Database.class.inc

```diff
@@ -500,16 +517,26 @@ class Database
         $err = $this->trackChanges($table, array(), $where, 'D');
 
         $this->_printQuery($query);
-        $this->affected = $this->_PDO->exec($query);
-        if ($this->affected === false) {
+        $result = $this->_PDO->exec($query);
+        if ($result === false) {
             throw new DatabaseException(
                 "Database DELETE did not execute successfully."
                 . print_r($this->_PDO->errorInfo(), true),
                 $query,
                 $where
             );
-
+        } elseif ($result === 0) {
+            if (DEBUG) {
+                fwrite(
+                    STDERR,
+                    "Database DELETE executed successfully "
+                    . "but 0 matched rows: " . $query. "\n"
+                );
+                return;
+            }
         }
+
+        $this->affected = $result;
```

Comment:
But 0 results affected isn't an error condition, let alone an exceptional one. It's just an execution of a query where nothing matched the where clause. That's a perfectly valid thing for an SQL statement to do.

A debug statement seems like the right level of warning to provide, but this->affected should still be updated since the last statement affected 0 rows.